src/transport/manager/mod.rs

```diff
@@ -254,57 +254,112 @@ pub struct TransportManager {
     opening_errors: HashMap<ConnectionId, Vec<(Multiaddr, DialError)>>,
 }
 
-impl TransportManager {
-    /// Create new [`crate::transport::manager::TransportManager`].
-    // TODO: https://github.com/paritytech/litep2p/issues/343 don't return handle here
-    pub fn new(
-        keypair: Keypair,
+pub struct TransportManagerBuilder {
+    keypair: Option<Keypair>,
+    supported_transports: HashSet<SupportedTransport>,
+    bandwidth_sink: Option<BandwidthSink>,
+    max_parallel_dials: usize,
+    connection_limits_config: limits::ConnectionLimitsConfig,
+}
+
+impl Default for TransportManagerBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TransportManagerBuilder {
+    /// Create new [`crate::transport::manager::TransportManagerBuilder`].
+    pub fn new() -> Self {
+        Self {
+            keypair: None,
+            supported_transports: HashSet::new(),
+            bandwidth_sink: None,
+            max_parallel_dials: MAX_PARALLEL_DIALS,
+            connection_limits_config: limits::ConnectionLimitsConfig::default(),
+        }
+    }
+
+    /// Set the keypair
+    pub fn with_keypair(&mut self, keypair: Keypair) -> &mut Self {
+        self.keypair = Some(keypair);
+        self
+    }
+
+    /// Set the supported transports
+    pub fn with_supported_transports(
+        &mut self,
         supported_transports: HashSet<SupportedTransport>,
-        bandwidth_sink: BandwidthSink,
-        max_parallel_dials: usize,
+    ) -> &mut Self {
+        self.supported_transports = supported_transports;
+        self
+    }
+
+    /// Set the bandwidth sink
+    pub fn with_bandwidth_sink(&mut self, bandwidth_sink: BandwidthSink) -> &mut Self {
+        self.bandwidth_sink = Some(bandwidth_sink);
+        self
+    }
+
+    /// Set the maximum parallel dials per peer
+    pub fn with_max_parallel_dials(&mut self, max_parrallel_dials: usize) -> &mut Self {
+        self.max_parallel_dials = max_parrallel_dials;
+        self
+    }
+
+    /// Set connection limits configuration.
+    pub fn with_connection_limits_config(
+        &mut self,
         connection_limits_config: limits::ConnectionLimitsConfig,
-    ) -> (Self, TransportManagerHandle) {
+    ) -> &mut Self {
+        self.connection_limits_config = connection_limits_config;
+        self
+    }
+
+    /// Build [`TransportManager`].
+    pub fn build(&mut self) -> TransportManager {
```

Comment:
This should take the ownership of builder:
```suggestion
    pub fn build(self) -> TransportManager {
```